### PR TITLE
Private communities

### PR DESCRIPTION
**Pull request by brianlovin**
**Status**
- [ ] WIP
- [x] Ready for review
- [ ] Needs testing

**Deploy after merge (delete what needn't be deployed)**
- api
- hyperion (frontend)
- athena
- vulcan
- hermes

**Run database migrations (delete if no migration was added)**
YES

WIP

### CODE REVIEW

File: api/models/usersCommunities.js [hunk 1]
```diff
@@ -342,13 +346,148 @@ const removeUsersCommunityMemberships = async (userId: string) => {
...
+    .then(result => {
+      if (result && result.length > 0) {
+        // if the result exists, it means the user has a previous relationship
+        // with this community - we handle blocked logic upstream in the mutation,
+        // so in this case we can just update the record to be pending
+
+        return db
+          .table('usersCommunities')
+          .getAll(userId, { index: 'userId' })
+          .filter({ communityId })
```
Comment (mxstbr) [inline api/models/usersCommunities.js]: We got a `userIdAndCommunityId` index, use that!

File: api/models/usersCommunities.js [hunk 2]
```diff
@@ -342,13 +346,148 @@ const removeUsersCommunityMemberships = async (userId: string) => {
...
+  trackQueue.add({
+    userId,
+    event: events.USER_CANCELED_REQUEST_TO_JOIN_COMMUNITY,
+    context: { communityId }
+  })
+
+  return db
+    .table('usersCommunities')
+    .getAll(communityId, { index: 'communityId' })
+    .filter({ userId })
```
Comment (mxstbr) [inline api/models/usersCommunities.js]: `userIdAndCommunityId` index

File: api/models/usersCommunities.js [hunk 3]
```diff
@@ -342,13 +346,148 @@ const removeUsersCommunityMemberships = async (userId: string) => {
...
+}
+
+export const approvePendingMemberInCommunity = async (
+  communityId: string,
+  userId: string
+): Promise<DBUsersCommunities> => {
+  return db
+    .table('usersCommunities')
+    .getAll(communityId, { index: 'communityId' })
+    .filter({ userId })
```
Comment (mxstbr) [inline api/models/usersCommunities.js]: `userIdAndCommunityId` index

File: api/models/usersCommunities.js [hunk 4]
```diff
@@ -342,13 +346,148 @@ const removeUsersCommunityMemberships = async (userId: string) => {
...
+};
+
+export const blockPendingMemberInCommunity = async (
+  communityId: string,
+  userId: string
+): Promise<DBUsersCommunities> => {
+  return db
+    .table('usersCommunities')
+    .getAll(communityId, { index: 'communityId' })
+    .filter({ userId })
```
Comment (mxstbr) [inline api/models/usersCommunities.js]: `userIdAndCommunityId` index

File: api/models/usersCommunities.js [hunk 5]
```diff
@@ -342,13 +346,148 @@ const removeUsersCommunityMemberships = async (userId: string) => {
...
 
   return Promise.all([...trackingPromises, removeMembershipsPromise]);
 };
 
+// prettier-ignore
+export const createPendingMemberInCommunity = async (communityId: string, userId: string): Promise<DBUsersCommunities> => {
+  return db
+    .table('usersCommunities')
+    .getAll(userId, { index: 'userId' })
+    .filter({ communityId })
```
Comment (mxstbr) [inline api/models/usersCommunities.js]: `userIdAndCommunityId` index

**Review by mxstbr — COMMENTED** [review]
Can we reuse the same infra for member approval?

Token join seems like a general solution for both private communities and member approval from afar.

**Comment by brianlovin** [comment]
> Token join seems like a general solution for both private communities and member approval from afar.

I thought the same thing and tried at first - it's frustrating because *most* of the logic is the same, but there's pockets of small differences in logic that would be more confusing to abstract around than to just handle these two one-off scenarios :)

> We got a userIdAndCommunityId index, use that!

Will fix!

**Comment by brianlovin** [comment]
@mxstbr this is ready for your code review and testing - ideally we can get this on alpha tomorrow and ship it mid-week :)

**Comment by brianlovin** [comment]
The failing e2e tests there all seem to be passing locally for me @mxstbr - when you review this any chance you could see if they are failing on your end?

**Review by mxstbr — COMMENTED** [review]
Skimmed over the code, overall that looks fine. A couple UI nitpicks:

- It's very very jarring that the "has requested to join your private community" notification icon is the person with the `+`. (the same as the "this person has joined your community" notification) Can we replace that `+` with a `?`?

screen shot 2018-05-21 at 11 52 50

- Why do I need to do two clicks to approve or block members? Can we just put the "Approve"/"Block"/"DM" buttons inline? Imagine if you just opened a community and hundreds of people are requesting to join...

screen shot 2018-05-21 at 11 53 00

- When I'm on the community view of a private community I've requested to join we should poll for that communities' data in the background. (ever 5-10s) Otherwise I could be staring at that for hours without realizing that I've been approved, that should immediately show the community once that happens.

Other thoughts:

- Is the system built to handle these two future cases?
  - Public communities that are approval-only
  - Join applications with a form that users who want to request to join fill out and answer questions, both for private and for public approval-only communities

**Review by mxstbr — DISMISSED** [review]
- The "should create a private community" e2e test also fails for me locally.

**Comment by brianlovin** [comment]
> The "should create a private community" e2e test also fails for me locally.

Fixed - i went over the character limit in the test name and didn't realize :P

> It's very very jarring that the "has requested to join your private community" notification icon is the person with the +. (the same as the "this person has joined your community" notification) Can we replace that + with a ??

Yeah, let me see if Bryn has an icon for that.

> Why do I need to do two clicks to approve or block members? Can we just put the "Approve"/"Block"/"DM" buttons inline? Imagine if you just opened a community and hundreds of people are requesting to join...

Agree, this is some technical debt we have on private channel requests as well.

> When I'm on the community view of a private community I've requested to join we should poll for that communities' data in the background. (ever 5-10s) Otherwise I could be staring at that for hours without realizing that I've been approved, that should immediately show the community once that happens.

Do you think people would sit there and wait? I'm not sure - the user gets both a notification and an email when their request was approved.

> Public communities that are approval-only

This does not design for that use case, although I think that can still be handled well in the `joinSettings` part of the `communitySettings` model.

> Join applications with a form that users who want to request to join fill out and answer questions, both for private and for public approval-only communities

This isn't designed for this either, but again I think that can be handled in the `communitySettings` model fairly well if we need it. The introduction text itself would be stored on `usersCommunities` records.

**Comment by brianlovin** [comment]
I've gone ahead and made the notification itself much more useful in these situations, to account for the technical debt of that dropdown in the community members settings view:

screen shot 2018-05-21 at 8 22 00 am

**Comment by brianlovin** [comment]
> I've gone ahead and made the notification itself much more useful in these situations, to account for the technical debt of that dropdown in the community members settings view:

Note: we should eventually do this for private channel requests, but we have a bunch of technical debt that we'd need to solve to make this a good implementation. Specifically, we need to have the same concept of `channelMember` as we do for `communityMember` :)

**Comment by brianlovin** [comment]
Hmm @orta - just a heads up, Peril is failing because:

screenshot 2018-05-21 09 00 58

However, that file definitely has flow enabled:
screenshot 2018-05-21 09 01 30

I've noticed this starting to happen quite a bit in the last week or two where it will flag files as not having flow enabled when they do.

cc @mxstbr

**Comment by brianlovin** [comment]
In instances like these, it would be awesome to have a way to manually re-trigger a peril run somehow :)

**Comment by brianlovin** [comment]
Additionally @orta, most of these files have flow enabled as well:

screenshot 2018-05-21 09 07 
17

I tried pushing up a new commit to trigger a rebuild but it insta-failed.

**Comment by brianlovin** [comment]
These test failures are caused by a memory error on CI :/

**Comment by orta** [comment]
I wonder if this is still about large PRs that cause the plugin to not act like expected, e.g. if it's hitting the GH request limits again peril will probably be returning an empty response which won't pass an includes check for `@flow`

**Comment by brianlovin** [comment]
I wish GH would just greenlight you to make infinite requests :P

**Comment by orta** [comment]
I wonder if the v4 GraphQL API could be abused for this, as that can now be used by graphQL apps, could batch through 10 files at a time or something.

**Pull request merged**